monty-31/src/dft/mod.rs

```diff
@@ -0,0 +1,296 @@
+///     - output is bit-reversed, as required for FRI.
+///
+/// Hence the only bit-reversal that needs to take place is on the input.
+///
+impl<MP: MontyParameters + FieldParameters + TwoAdicData> TwoAdicSubgroupDft<MontyField31<MP>>
+    for RecursiveDft<MontyField31<MP>>
+{
+    type Evaluations = BitReversedMatrixView<RowMajorMatrix<MontyField31<MP>>>;
+
+    #[instrument(skip_all, fields(dims = %mat.dimensions(), added_bits))]
+    fn dft_batch(&self, mut mat: RowMajorMatrix<MontyField31<MP>>) -> Self::Evaluations
+    where
+        MP: MontyParameters + FieldParameters + TwoAdicData,
+    {
+        let nrows = mat.height();
+        let ncols = mat.width();
+        if nrows <= 1 {
+            return mat.bit_reverse_rows();
+        }
+
+        let mut scratch = debug_span!("allocate scratch space")
+            .in_scope(|| RowMajorMatrix::default(nrows, ncols));
+
+        self.update_twiddles(nrows);
+        let twiddles = self.twiddles.borrow().clone();
+
+        // transpose input
+        debug_span!("pre-transpose", nrows, ncols)
+            .in_scope(|| transpose::transpose(&mat.values, &mut scratch.values, ncols, nrows));
+
+        debug_span!("dft batch", n_dfts = ncols, fft_len = nrows)
+            .in_scope(|| Self::decimation_in_freq_dft(&mut scratch.values, nrows, &twiddles));
+
+        // transpose output
+        debug_span!("post-transpose", nrows = ncols, ncols = nrows)
+            .in_scope(|| transpose::transpose(&scratch.values, &mut mat.values, nrows, ncols));
+
+        mat.bit_reverse_rows()
+    }
+
+    #[instrument(skip_all, fields(dims = %mat.dimensions(), added_bits))]
+    fn idft_batch(&self, mat: RowMajorMatrix<MontyField31<MP>>) -> RowMajorMatrix<MontyField31<MP>>
+    where
+        MP: MontyParameters + FieldParameters + TwoAdicData,
+    {
+        let nrows = mat.height();
+        let ncols = mat.width();
+        if nrows <= 1 {
+            return mat;
+        }
+
+        let mut scratch = debug_span!("allocate scratch space")
+            .in_scope(|| RowMajorMatrix::default(nrows, ncols));
+
+        // TODO: Use faster bit-reversal algo
+        let mut mat =
+            debug_span!("initial bitrev").in_scope(|| mat.bit_reverse_rows().to_row_major_matrix());
+
+        self.update_inv_twiddles(nrows);
+        let inv_twiddles = self.inv_twiddles.borrow().clone();
+
+        // transpose input
+        debug_span!("pre-transpose", nrows, ncols)
+            .in_scope(|| transpose::transpose(&mat.values, &mut scratch.values, ncols, nrows));
+
+        debug_span!("idft", n_dfts = ncols, fft_len = nrows)
+            .in_scope(|| Self::decimation_in_time_dft(&mut scratch.values, nrows, &inv_twiddles));
+
+        // transpose output
+        debug_span!("post-transpose", nrows = ncols, ncols = nrows)
+            .in_scope(|| transpose::transpose(&scratch.values, &mut mat.values, nrows, ncols));
+
+        let inv_len = MontyField31::from_canonical_usize(nrows).inverse();
```

Comment:
This is probably not super expensive as it's just one inverse but this inverse is unnecessary. We should make a const fn in monty-31 utils which returns 2^{-n} as, provided n < 32 (which it always is) the monty form of 2^{-n} is precisely 2^{32 - n} which (for n =/= 0, 1) is already in [0, ..., P).